File: core/java/src/main/java/hara/lang/protocol/IWork.java
package hara.lang.protocol;

import hara.lang.declaration.HaraAvailability;
import hara.lang.declaration.HaraMethod;
import hara.lang.declaration.HaraProtocolBinding;

/** A stable, side-effect-free description of executable work. */
@HaraProtocolBinding(
    namespace = "std.protocol.iwork",
    name = "IWork",
    availability = HaraAvailability.CAPABILITY_GATED,
    capability = "native-runtime-protocols")
public interface IWork {
  @HaraMethod(value = "work-spec", arity = 1)
  Object workSpec();
}

File: core/rust/src/lang/protocol/iwork.rs
use hara_protocol_macros::hara_protocol;

#[hara_protocol(
    namespace = "std.protocol.iwork",
    name = "IWork",
    availability = "capability-gated",
    capability = "native-runtime-protocols"
)]
pub trait IWork {
    type Spec;

    #[hara_method(value = "work-spec", arity = 1)]
    fn work_spec(&self) -> Self::Spec;
}
